lib/api/CDataFrameAnalysisRunner.cc

```diff
@@ -73,11 +73,15 @@ void CDataFrameAnalysisRunner::computeAndSaveExecutionStrategy() {
     LOG_TRACE(<< "number partitions = " << m_NumberPartitions);
 
     if (memoryUsage > memoryLimit) {
+        auto roundMb = [](std::size_t memory) {
+            return 0.01 * static_cast<double>((100 * memory) / (1024 * 1024));
```

Comment:
What is the benefit of switching this code to use floating point rather than doing the calculation using integer arithmetic?

Also, as it stands the lambda doesn't do the rounding, as that's done by the `std::ceil` that's separate.

If it's going to use floating point then can't the formula simply be `return static_cast<double>(memory) / static_cast<double>(1024 * 1024)`?

But I don't really see a need for this whole change.

Reply:
I changed it because if the actual limit is small the reported limit could easily be 0MB: for example you could get weird messages like `"Input error: memory limit 0MB is too low to perform analysis. You need to give the process at least 3MB, but preferably more."` The `0.01 * 100 * ` is just to get it to 2 d.p.

Reply:
> The `0.01 * 100 *` is just to get it to 2 d.p.

But the `std::ceil` in the log macro wipes out the decimals.

> if the actual limit is small the reported limit could easily be 0MB

The old formula was rounding up to the next MB, so 1 byte was and still is reported as 1MB.  The combination of converting to MB rounded to 2 d.p. and then taking the `ceil` is equivalent to the old formula.

Reply:
... so the round in the name applies refers to the rounding to 2d.p. @droberts195 are you happy with this explanation?  Do you have any suggestions for a better alternative?

Reply:
> But the std::ceil in the log macro wipes out the decimals.
...
  The old formula was rounding up to the next MB, so 1 byte was and still is reported as 1MB. The combination of converting to MB rounded to 2 d.p. and then taking the ceil is equivalent to the old formula.

This isn't quite right. The original code was (rounding down):
```
HANDLE_FATAL(<< "Input error: memory limit " << memoryLimit / (1024 * 1024)
```
for reporting the memory limit. There are now two calls: one rounds to 2d.p. (for current limit) one rounds to up to nearest MB (for required memory limit). We could make both round up. My rationale for reporting the actual limit more accurately is that I thought people might find it confusing if this is significantly different from what they know they've set. Either way I think we should avoid the 0MB.

Reply:
Sorry, I missed that that lambda was used in two places.

The Java code already enforces a minimum of 1MB for the model memory limit.  But it's true you could still configure it to something like 1.43MB.  Since we're reporting the limit to 2 d.p. maybe we should just get rid of the `std::ceil` and also report the requirement to 2 d.p.?

(It makes me think we should have just done what we did in anomaly detection and enforce that the model memory limit was a whole number of MB, as I'm sure 99.9% of people will set it this way.)

Reply:
The reason for leaving the round up was to give us some margin given that the required memory is the absolute minimum we can get away with as a limit. We need some pad, but maybe this could be a fixed pad and still report to 2d.p. The one observation, I have is it seems kind of odd to me to be so precise in a required memory statement (given that this includes various things we can only approximately estimate).

> It makes me think we should have just done what we did in anomaly detection and enforce that the model memory limit was a whole number of MB, as I'm sure 99.9% of people will set it this way

This is probably right. The one downside is that it can be wasteful, say if you don't have much data for training a prediction model, where you can get away with significantly less than 1MB. Given where we are now, I'm not sure it is worthwhile making that change.